### PR TITLE
Add preview tabs

### PR DESCRIPTION
**Pull request by bennetbo**
This PR implements the preview tabs feature from VSCode.
More details and thanks for the head start of the implementation here #6782.

Here is what I have observed from using the vscode implementation ([x] -> already implemented):
- [x] Single click on project file opens tab as preview
- [x] Double click on item in project panel opens tab as permanent
- [x] Double click on the tab makes it permanent
- [x] Navigating away from the tab makes the tab permanent and the new tab is shown as preview (e.g. GoToReference)
- [x] Existing preview tab is reused when opening a new tab
- [x] Dragging tab to the same/another panel makes the tab permanent
- [x] Opening a tab from the file finder makes the tab permanent
- [x] Editing a preview tab will make the tab permanent
- [x] Using the space key in the project panel opens the tab as preview
- [x] Handle navigation history correctly (restore a preview tab as preview as well)
- [x] Restore preview tabs after restarting
- [x] Support opening files from file finder in preview mode (vscode: "Enable Preview From Quick Open")
 
I need to do some more testing of the vscode implementation, there might be other behaviors/workflows which im not aware of that open an item as preview/make them permanent.

Showcase:

https://github.com/zed-industries/zed/assets/53836821/9be16515-c740-4905-bea1-88871112ef86

TODOs
- [x] Provide `enable_preview_tabs` setting
- [x] Write some tests
- [x] How should we handle this in collaboration mode (have not tested the behavior so far)
- [x] Keyboard driven usage (probably need workspace commands)
- [x] Register `TogglePreviewTab` only when setting enabled?
- [x] Render preview tabs in tab switcher as italic
- [x] Render preview tabs in image viewer as italic
- [x] Should this be enabled by default (it is the default behavior in VSCode)?
- [x] Docs

Future improvements (out of scope for now):
- Support preview mode for find all references and possibly other multibuffers
- Support VSCode: "Enable Preview From Code Navigation"

Release Notes:

- Added preview tabs ([#4922](https://github.com/zed-industries/zed/issues/4922)).

### CODE REVIEW

**Comment by 0x2CA** [comment]
The preview tabs I think makes #8013 possible

**Comment by baldwindavid** [comment]
@bennetbo Thank you for working on this feature. I'm curious if this will (or provides the underlying architecture) to support keyboard-driven preview? I don't see a command like `project_panel::Preview` in the changes so assuming that might be a future enhancement.

**Comment by bennetbo** [comment]
> @bennetbo Thank you for working on this feature. I'm curious if this will (or provides the underlying architecture) to support keyboard-driven preview? I don't see a command like `project_panel::Preview` in the changes so assuming that might be a future enhancement.

@baldwindavid sounds like a good idea. There now is an action `pane: toggle preview tab`, which lets you transform a "normal" tab into a preview tab and a preview tab into a "normal" tab.

Combined with using `space` to open items in the project panel and a key bind to focusing the project panel this works quite nicely as you can see here:

https://github.com/zed-industries/zed/assets/53836821/4d5db8d8-277b-478f-a65f-394b39a18d4e

What do you think? Im glad to get input on figuring out the best keyboard way to use preview tabs (Im also not sure if VSCode handles this nicely?).

**Comment by baldwindavid** [comment]
@bennetbo Seems like this has a lot of the vscode interaction covered. One difference here with vscode is that in vscode you can press "space" to preview files, but without moving the cursor to the file. This allows previewing files quickly without having to move back and forth between the project panel and editor. In fact, it appears the "single-click" behavior in this PR does work this way. I think it's not the same with "space" because it's just [defaulted](https://github.com/zed-industries/zed/blob/main/assets/keymaps/default-macos.json#L575) as "project_panel::Open". I guess I still wonder if a "project_panel::Preview" command could be introduced and then either default to "space" running that command if preview tabs are enabled or just let others map as needed.

In practice, I haven't been using the project panel at all in favor of using a commandline center terminal https://github.com/zed-industries/zed/discussions/9067. With these preview tabs I could see using it for project exploration though. If there were a command like "project_panel::Open" I'd likely map the up and down keys (j and k) to both move to the next file AND preview for live previews while navigating through the file tree similar to what I do with the yazi file explorer.

**Comment by baldwindavid** [comment]
> I'd likely map the up and down keys (j and k) to both move to the next file AND preview for live previews while navigating through the file tree similar to what I do with the yazi file explorer.

By this I mean mapping to a similar effect of this...

```json
{
  "context": "ProjectPanel && not_editing",
  "bindings": {
    "j": ["menu::SelectNext", "project_panel::Preview"], 
    "k": ["menu::SelectPrevious", "project_panel::Preview"], 
  }
},
```

...though think it would have to be kludged leveraging "workspace::SendKeystrokes".

**Comment by bennetbo** [comment]
@baldwindavid thanks for the great suggestions.
`project_panel::Open` will now open a preview tab if the preview tab feature is enabled, otherwise it will open a "permanent" tab. Also there is a new command `project_panel::OpenPermanent`, that will always open a tab as "permanent". 

E.g. I have `project_panel::Open` mapped to `space` and `project_panel::OpenPermanent` mapped to `shift-space`, allowing me to open preview/permanent tabs in a keyboard driven way.

Also changed the focus behavior when opening files from the project panel, zed now keeps the project panel focused (like VSCode does as well).

**Comment by baldwindavid** [comment]
Very cool. I think I might expect `project_panel::OpenPermanent` to both open a permanent tab AND move focus to that tab. For example, in vscode pressing `l` opens the file and focuses it, which I think might be similar to your concept of `OpenPermanent`. Otherwise, we need to hit another key toggle focus in that case (`projectPanel::ToggleFocus`).

Other than that, it probably just comes down to semantics of whether "open" means "previewing and focusing" or just "previewing"; you've chosen "previewing" here. I might have naively thought the opposite (`project_panel::Preview` for preview and `project_panel::Open` for open/permanent), but I think best to get feedback from the maintainers now or soon as I'm just one guy and don't even use the project panel as of now 😸 . Regardless, the core behavior is really nice here.

**Comment by bennetbo** [comment]
> Very cool. I think I might expect `project_panel::OpenPermanent` to both open a permanent tab AND move focus to that tab. For example, in vscode pressing `l` opens the file and focuses it, which I think might be similar to your concept of `OpenPermanent`. Otherwise, we need to hit another key toggle focus in that case (`projectPanel::ToggleFocus`).
> 
> Other than that, it probably just comes down to semantics of whether "open" means "previewing and focusing" or just "previewing"; you've chosen "previewing" here. I might have naively thought the opposite (`project_panel::Preview` for preview and `project_panel::Open` for open/permanent), but I think best to get feedback from the maintainers now or soon as I'm just one guy and don't even use the project panel as of now 😸 . Regardless, the core behavior is really nice here.

I did another small change, `project_panel::OpenPermanent` will now transfer focus to the editor, whereas `project_panel::Open` will keep the project panel focused. I think that will allow better flexibility when customizing the keybindings and also provide an easy way to mimic the old behavior of the project panel (disable preview tabs settings and rebind `Open` to `OpenPermanent`).

The reason why I did not go for the `Preview`/`Open` naming here, is that `Preview` will not open a tab as preview when the setting is disabled. Therefore I decided to keep `Open` and that will either open a "permanent" tab when the setting is disabled, or a preview tab if the setting is enabled. Then there is also `OpenPermanent` which will focus the editor and will always open a "permanent" tab.

**Comment by 0x2CA** [comment]
in vim ,Keymap `"space f p": "pane::TogglePreviewTab"`,Interface not updated.

https://github.com/zed-industries/zed/assets/19868292/0782177f-991c-4663-88ac-70bbf3a1ce9c

**Comment by bennetbo** [comment]
> in vim ,Keymap `"space f p": "pane::TogglePreviewTab"`,Interface not updated.
> 
>  2024-04-12.11.18.40.mov

Hey @0x2CA,
I am not a vim user, so I don't know what default keybindings vim users would expect. Is there any other related keybindings that are missing when using preview tabs? If so we should put those in.

E.g. in the vim keymap there are four ways to open a file from the project panel right now, should any of those become `project_panel::OpenPermanent`?
```json
"enter": "project_panel::Open",
"o": "project_panel::Open",
"t": "project_panel::Open",
"v": "project_panel::Open",
```

**Comment by baldwindavid** [comment]
@bennetbo I'd preface by saying that I don't feel that strongly about any of this. We're trying to balance concepts from Zed, VSCode, Vim, and Netrw so don't think there is a "right" answer. Also, goes without saying, devs can remap to their heart's content. :)

At any rate, the reason "space f p" will not work is because "space" is defaulted to open a file in `default-macos.json`...

```json
{
  "context": "ProjectPanel && not_editing",
  "bindings": {
    "space": "project_panel::Open"
  }
},
```

This isn't really a vim/netrw thing, but "space" in the VSCode panel previews and I think it's an okay default in Zed too, even for vim mode. 

Note that there "is" the concept of previewing in netrw and it's bound to "p" so think that would be a good default to add. Other than that, I tend to think that all the others insinuate opening permanently.

In other words, in addition to keeping the "space" keybinding as-is...

```json
"enter": "project_panel::OpenPermanent",
"o": "project_panel::OpenPermanent",
"t": "project_panel::OpenPermanent",
"v": "project_panel::OpenPermanent",
"p": "project_panel::Open",
```

VSCode also effectively opens permanently via "l" (the key to move "right" in vim). We've got that mapped to `project_panel::ExpandSelectedEntry` currently as a way to expand directories in the panel. I think, ideally, we would have a command that could expand a directory OR open a file permanently and then default "l" to perform that command. VSCode does that currently.

@ConradIrwin may have some opinions on this too.

**Comment by ConradIrwin** [comment]
:+1: to what you said @baldwindavid https://github.com/zed-industries/zed/pull/10492

**Comment by baldwindavid** [comment]
Really nice work on all this, @bennetbo ! Congrats on getting it across the line.

**Comment by edtorba** [comment]
Top job 👍 Definitely a quality-of-life feature that many wanted
